src/modules/hip/kernel/warp_perspective.cpp

```diff
@@ -1,137 +1,184 @@
-    unsigned long src_pixIdx, dst_pixIdx;
-    src_pixIdx = source_batch_index[id_z] +
-                 (k + l * max_source_width[id_z]) * plnpkdindex;
-    dst_pixIdx = dest_batch_index[id_z] +
-                 (id_x + id_y * max_dest_width[id_z]) * plnpkdindex;
-    for (indextmp = 0; indextmp < channel; indextmp++) {
-      dstPtr[dst_pixIdx] = srcPtr[src_pixIdx];
-      src_pixIdx += source_inc[id_z];
-      dst_pixIdx += dest_inc[id_z];
+extern "C" __global__ void warp_perspective_batch(unsigned char *srcPtr,
+                                                  unsigned char *dstPtr,
+                                                  float *perspective,
+                                                  unsigned int *source_height,
+                                                  unsigned int *source_width,
+                                                  unsigned int *dest_height,
+                                                  unsigned int *dest_width,
+                                                  unsigned int *xroi_begin,
+                                                  unsigned int *xroi_end,
+                                                  unsigned int *yroi_begin,
+                                                  unsigned int *yroi_end,
+                                                  unsigned int *max_source_width,
+                                                  unsigned int *max_dest_width,
+                                                  unsigned long long *source_batch_index,
+                                                  unsigned long long *dest_batch_index,
+                                                  const unsigned int channel,
+                                                  unsigned int *source_inc, // use width * height for pln and 1 for pkd
+                                                  unsigned int *dest_inc,
+                                                  const int plnpkdindex) // use 1 pln 3 for pkd
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    if (id_x >= dest_width[id_z] || id_y >= dest_height[id_z])
+    {
+        return;
+    }
+
+    int indextmp = 0;
+    int xc = id_x - (dest_width[id_z] >> 1);
+    int yc = id_y - (dest_height[id_z] >> 1);
+    int perspective_index = id_z * 9;
+
+    float k_float = (float)((perspective[perspective_index + 0] * xc) + (perspective[perspective_index + 1] * yc)) + perspective[perspective_index + 2];
+    float l_float = (float)((perspective[perspective_index + 3] * xc) + (perspective[perspective_index + 4] * yc)) + perspective[perspective_index + 5];
+    float m_float = (float)((perspective[perspective_index + 6] * xc) + (perspective[perspective_index + 7] * yc)) + perspective[perspective_index + 8];
+
+    int k = (int) ((k_float / m_float) + (source_width[id_z] >> 1));
```

Comment:
need to revisit this to avoid this division

Reply:
noted